File: IGImageCropper/Classes/IGCropViewController.m
//
//  IGCropViewController.m
//  IGImageCropper
//
//  Created by arshad tp on 15/09/21.
//

#import "IGCropViewController.h"
#import "UIImage+ImageLoad.h"


typedef NS_ENUM(NSInteger, ZoomScale) {
    ZoomScaleMinimum,
    ZoomScaleSquare,
};

static NSString * const _Nonnull FrameworkBundleId = @"org.cocoapods.IGImageCropper";

@interface IGCropViewController () {
    
    ZoomScale activeZoomScale;
    CGFloat initialScale;

}

@property (nonatomic) CGFloat minimumPortraitZoomScale;
@property (nonatomic) CGFloat minimumLandScapeZoomScale;
@property (nonatomic, strong) UIImage *image;
@property (weak, nonatomic) IBOutlet UIView *scrollViewContainer;
@property (unsafe_unretained, nonatomic) IBOutlet UIScrollView *scrollView;
@property (unsafe_unretained, nonatomic) IBOutlet UIButton *adjustZoomScaleButton;
@property (strong, nonatomic) IBOutletCollection(UIView) NSArray *cropMask;
@property (unsafe_unretained, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *cancelButton;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *chooseButton;
@property (unsafe_unretained, nonatomic) IBOutlet NSLayoutConstraint *imageViewHeightConstraint;
@property (unsafe_unretained, nonatomic) IBOutlet NSLayoutConstraint *imageViewWidthConstraint;

@end

@implementation IGCropViewController


- (instancetype) initWithImage: (UIImage *) image minimumPortraitZoomScale: (CGFloat) minimumPortraitZoomScale minimumLandScapeZoomScale: (CGFloat) minimumLandScapeZoomScale  {
    
    self.minimumLandScapeZoomScale = minimumLandScapeZoomScale;
    self.minimumPortraitZoomScale = minimumPortraitZoomScale;
    self.image = image;
    initialScale = 1.0;
    activeZoomScale = ZoomScaleSquare;
    self = [super initWithNibName:@"IGCropViewController" bundle:[NSBundle bundleWithIdentifier:FrameworkBundleId]];
    return self;
    
}


#pragma mark - Life Cycle Methods
- (void)viewDidLoad {
    [super viewDidLoad];
    [self configureView];
    [self setImageToCrop];
}


#pragma mark - Util Methods

- (BOOL) isPortraitImage {
    return _image.size.width <= _image.size.height;
}

- (CGFloat) minZoomScale {
    if ([self isPortraitImage]) {
        return _minimumPortraitZoomScale;
    }
    return  _minimumLandScapeZoomScale;
}

- (UIImage *) cropImage {
    
    CGFloat zoomScale = _scrollView.zoomScale;
    CGFloat scale = 1/zoomScale * 1/initialScale;
    
    CGFloat x =  _scrollView.contentOffset.x * scale;
    CGFloat y = _scrollView.contentOffset.y * scale;
    CGFloat width = _scrollView.frame.size.width  * scale;
    CGFloat height = _scrollView.frame.size.height * scale;
    
    
    CGImageRef imageRef = CGImageCreateWithImageInRect(_imageView.image.CGImage, CGRectMake(x, y, width, height));
    
    UIImage *croppedImage = [[UIImage alloc] initWithCGImage:imageRef];

    return croppedImage;
}

- (void) animateMaskViewToAlpha: (CGFloat) alpha {
    
    __weak typeof(self) weakSelf = self;
    [UIView animateWithDuration:1.5 animations:^{
        for (UIView *view in weakSelf.cropMask) {
            view.alpha = alpha;
        }
    }];
}

#pragma mark - Configs
- (void) configureView {
    
    [_adjustZoomScaleButton setImage:[UIImage bundledImageNamed:@"shrink" bundleId:FrameworkBundleId] forState:UIControlStateNormal];
    [_adjustZoomScaleButton setImage:[UIImage bundledImageNamed:@"expand" bundleId:FrameworkBundleId] forState:UIControlStateNormal];
    if (_chooseButtonTitle != NULL) {
        [_chooseButton setTitle:_chooseButtonTitle];
    }
    if (_cancelButtonTitle != NULL) {
        [_cancelButton setTitle:_cancelButtonTitle];
    }
    
//    _scrollViewContainer.layer.borderColor = (__bridge CGColorRef _Nullable)([UIColor colorWithRed:128/255 green:128/255 blue:128/255 alpha:1]);
//    _scrollViewContainer.layer.borderWidth = 10.0;
}

- (void) setImageToCrop {
    
    _imageView.image = _image;
    
    if ([self isPortraitImage]) {
        // For portait image make width maximum as screen width and adjust the height accordingly
        CGFloat width = UIScreen.mainScreen.bounds.size.width;
        initialScale = width/_image.size.width;

        _imageViewWidthConstraint.constant = width;
        _imageViewHeightConstraint.constant = _image.size.height * initialScale;
    } else {
        
        // For portait image make height maximum as screen width and adjust the width accordingly
        CGFloat height = UIScreen.mainScreen.bounds.size.width; // first showing in square crop where width = screen width
        initialScale = height/_image.size.height;
        
        _imageViewWidthConstraint.constant = _image.size.width * initialScale;
        _imageViewHeightConstraint.constant = height;
        
    }
    
    _scrollView.minimumZoomScale = [self minZoomScale];
    _scrollView.maximumZoomScale = 5.0;
    _scrollView.contentSize = _imageView.frame.size;
    _scrollView.delegate = self;
    
}

#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}

#pragma mark - IBActions

- (IBAction)adjustZoomScaleAction:(id)sender {
    
    if (activeZoomScale == ZoomScaleSquare) {
        
        activeZoomScale = ZoomScaleMinimum;
        [_adjustZoomScaleButton setSelected: YES];
        [_scrollView setZoomScale:[self minZoomScale] animated:YES];
    } else {
        
        activeZoomScale = ZoomScaleSquare;
        [_adjustZoomScaleButton setSelected: NO];

        [_scrollView setZoomScale:1 animated:YES];
    }
}

- (IBAction)cancelAction:(id)sender {
    
    [self.delegate cropViewController:self didFinishCancelled:YES];
}
- (IBAction)chooseButtonAction:(id)sender {
    
   UIImage *croppedImage = [self cropImage];
    [self.delegate cropViewController:self didCropToImage:croppedImage withRect:CGRectMake(0, 0, croppedImage.size.width, croppedImage.size.height)];
}


#pragma mark - ScrollView Delegates

- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
    
    return _imageView;
}

- (void)scrollViewDidZoom:(UIScrollView *)scrollView {
    
    if ([self isPortraitImage]) {
        
        CGFloat h =  (UIScreen.mainScreen.bounds.size.width - _imageView.frame.size.width) / 2;
        _scrollView.contentInset = UIEdgeInsetsMake(0, h < 0 ? 0 : h, 0, 0);
    } else {
        
        CGFloat v =  (UIScreen.mainScreen.bounds.size.width - _imageView.frame.size.height) / 2;
        _scrollView.contentInset = UIEdgeInsetsMake(v < 0 ? 0 : v, 0, 0, 0);

    }
}

- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView {
    
    [self animateMaskViewToAlpha:0.4];
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate {
    
    [self animateMaskViewToAlpha:1.0];
}

@end

File: IGImageCropper/Classes/IGCropViewController.swift
//
//  IGCropViewController.swift
//  Cropper
//
//  Created by arshad tp on 14/09/21.
//

import UIKit

public class CropInfo: NSObject {
    let minimumPortraitZoomScale: CGFloat = 4/5
    let minimumLandScapeZoomScale: CGFloat = 1/1.91
}

let frameworkBundleID =  "org.cocoapods.IGImageCropper"
public class IGCropViewController: UIViewController {
    
    enum ZoomScale {
        case minimum
        case square
    }
    
    let image: UIImage
    let cropInfo: CropInfo
    
    var activeZoomScale: ZoomScale = .square
    
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var scrollView: UIScrollView!
    @IBOutlet weak var imageViewHeightContraint: NSLayoutConstraint!
    @IBOutlet weak var imageViewWidthConstraint: NSLayoutConstraint!
    
    @IBOutlet weak var adjustZoomScaleButton: UIButton!
    @IBOutlet var cropMask: [UIView]!
    var initialScale: CGFloat = 1.0
    var isPortraitImage: Bool {
        return image.size.width <= image.size.height
    }
    
    var minZoomScale: CGFloat  {
        if (isPortraitImage) {
            return cropInfo.minimumPortraitZoomScale
        }
        return cropInfo.minimumLandScapeZoomScale
    }
    
    public init(image: UIImage, cropInfo: CropInfo?) {
        if let cropInfo = cropInfo {
            self.cropInfo = cropInfo
        } else {
            self.cropInfo = CropInfo()
        }
        self.image = image
        let bundle = Bundle(identifier: frameworkBundleID)

        super.init(nibName: "IGCropViewController", bundle: bundle)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    func setImageToCrop(image:UIImage){
        
        imageView.image = image
        if (isPortraitImage) {
            // For portait image make width maximum as screen width and adjust the height accordingly
            let width = UIScreen.main.bounds.width
            initialScale = width/image.size.width
            
            imageViewWidthConstraint.constant = width
            imageViewHeightContraint.constant = image.size.height * initialScale
            
        } else {
            // For portait image make height maximum as screen width and adjust the width accordingly
            let height = UIScreen.main.bounds.width // first showing in square crop where width = screen width
            initialScale = height/image.size.height
            
            imageViewWidthConstraint.constant = image.size.width * initialScale
            imageViewHeightContraint.constant = height
        }
        
        
        
        scrollView.minimumZoomScale = minZoomScale
        scrollView.maximumZoomScale = 5.0
        scrollView.contentSize = imageView.frame.size
        scrollView.delegate = self
    }
    public override func viewDidLoad() {
        super.viewDidLoad()
        
        print("image ", UIImage.bundledImage(named: "shrink"))
        adjustZoomScaleButton.setImage(UIImage.bundledImage(named: "shrink"), for: .normal)
        adjustZoomScaleButton.setImage(UIImage.bundledImage(named: "expand"), for: .selected)
        setImageToCrop(image: image)
        // Do any additional setup after loading the view.
    }
    
    @IBAction func cancelAction(_ sender: Any) {
    }
    
    @IBAction func chooseButtonAction(_ sender: Any) {
        
        cropImage()
    }
    
    @IBAction func adjustZoomScaleAction(_ sender: Any) {
        
        if (activeZoomScale == .square) {
            activeZoomScale = .minimum
            adjustZoomScaleButton.isSelected = true
            scrollView.setZoomScale(minZoomScale, animated: true)
        } else {
            activeZoomScale = .square
            adjustZoomScaleButton.isSelected = false
            scrollView.setZoomScale(1, animated: true)

        }
    }
    
    private func cropImage() {
        
        let zoomScale = scrollView.zoomScale
        let scale:CGFloat = 1/zoomScale * 1/initialScale
        
        let x:CGFloat =  scrollView.contentOffset.x * scale
        let y:CGFloat = scrollView.contentOffset.y * scale
        let width:CGFloat = scrollView.frame.size.width  * scale
        let height:CGFloat = scrollView.frame.size.height * scale
        
        print("crop rect ", CGRect(x: x, y: y, width: width, height: height))
        let croppedCGImage = imageView.image?.cgImage?.cropping(to: CGRect(x: x, y: y, width: width, height: height))
        let croppedImage = UIImage(cgImage: croppedCGImage!)
        
        print("scroll view ==" , scrollView.frame.size)
        print("result image size === ", croppedImage.size)
//        saveImage(image: croppedImage)
    }
//    func saveImage(image: UIImage) -> Bool {
//        guard let data = image.UIImageJPEGRepresentation(compressionQuality: 1) ?? image.pngData() else {
//            return false
//        }
//        guard let directory = try? FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false) as NSURL else {
//            return false
//        }
//        do {
//            try data.write(to: directory.appendingPathComponent("fileName.png")!)
//            print("path == ", directory.path)
//            return true
//        } catch {
//            print(error.localizedDescription)
//            return false
//        }
//    }
    
    private  func animateMaskView(alpha: CGFloat) {
        UIView.animate(withDuration: 1.5) { [weak self] in
            self?.cropMask.forEach { (view) in
                view.alpha = alpha
            }
        }
    }
}


extension IGCropViewController: UIScrollViewDelegate {
    
    public func viewForZooming(in scrollView: UIScrollView) -> UIView? {
        
        return imageView
    }
    
    public func scrollViewDidZoom(_ scrollView: UIScrollView) {
        if (isPortraitImage) {
            let h = (UIScreen.main.bounds.width - imageView.frame.width)/2
            scrollView.contentInset = UIEdgeInsets.init(top: 0, left: h < 0 ? 0 : h, bottom: 0, right: 0)
        } else {
            let v = (UIScreen.main.bounds.width - imageView.frame.height)/2
            scrollView.contentInset = UIEdgeInsets.init(top: v < 0 ? 0 : v, left: 0, bottom: 0, right: 0)
        }
        
        
    }
    
    
    
    public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        animateMaskView(alpha: 0.4)
    }
    
    public func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        animateMaskView(alpha: 1.0)
        
    }
    
    
}

extension UIImage {
    class func bundledImage(named: String) -> UIImage? {
        let image = UIImage(named: named)
        if image == nil {
            let bundle = Bundle(identifier: "org.cocoapods.IGImageCropper")

            return UIImage(named: named, in: bundle, compatibleWith: nil)
        } // Replace MyBasePodClass with yours
        return image
    }
}
